In [79]:
import pandas as pd

In [80]:
df = pd.read_csv('../data/pokemon_data.csv')
df.head(10)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False


In [81]:
df[df['HP'] > df['Attack']].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
13,10,Caterpie,Bug,NaN,45,30,35,20,20,45,1,False
14,11,Metapod,Bug,NaN,50,20,55,25,25,30,1,False
15,12,Butterfree,Bug,Flying,60,45,50,90,80,70,1,False
16,13,Weedle,Bug,Poison,40,35,30,20,20,50,1,False
17,14,Kakuna,Bug,Poison,45,25,50,25,25,35,1,False


In [82]:
## filter only with one type( no type 2)

df[df['Type 2'].isnull()].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False
10,8,Wartortle,Water,NaN,59,63,80,65,80,58,1,False
11,9,Blastoise,Water,NaN,79,83,100,85,105,78,1,False


# with and without loc

Your two lines

## A) Boolean filtering (most common)

df[df['Type 1'] == 'Grass'].head(5)

## B) .loc version

df.loc[df['Type 1'] == 'Grass'].head(5)

more explicit pandas method

label-based indexer

allows row + column selection together



## Key answer first

    In this case, both return the same result

    BUT the way they work internally is different.

| Feature                   | `df[...]` | `df.loc[...]`  |
| ------------------------- | --------- | -------------- |
| readability               | simpler   | clearer intent |
| column selection          | limited   | powerful       |
| row + column filtering    | no        | yes            |
| recommended in interviews | yes       | YES (better)   |


# Important insight
This works:
```python
df[df['Type 1'] == 'Grass']
```
But this is more powerful:
```python
df.loc[df['Type 1'] == 'Grass', ['Name', 'HP']]
```
## .loc = full control

Why .loc is preferred in serious code

Because it avoids ambiguity:

df[...] → could be row or column confusion

df.loc → explicitly row + column selector

======================================================


## One important subtle difference

If you later do assignment:
```python
df.loc[df['Type 1'] == 'Grass', 'HP'] = 0
```
It sets the HP of all Grass-type Pokémon to 0

✔ modifies safely

But:
```python
df[df['Type 1'] == 'Grass']['HP'] = 0
```
may fail (chained indexing problem)

# Final rule

.loc is safer and more powerful

df[...] is shorter but limited

===========================================================
# To change multiple columns in pandas, you just pass a list of column names instead of a single column.

## Basic idea
```python
df.loc[rows, [col1, col2, col3]] = value
```
Example 1: change multiple columns to same value

```python
df.loc[df['Type 1'] == 'Grass', ['HP', 'Attack']] = 0
```
sets HP and Attack to 0 for Grass Pokémon

In [83]:
df.loc[df['Type 1'] ==  'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [84]:
df[df['Type 1'] == 'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [85]:
df.loc[:,'Type 1']

0        Grass
1        Grass
2        Grass
3        Grass
4         Fire
        ...   
795       Rock
796       Rock
797    Psychic
798    Psychic
799       Fire
Name: Type 1, Length: 800, dtype: str

In [86]:
df.loc[df['Type 1'] == 'Grass', 'HP']

0       45
1       60
2       80
3       80
48      45
      ... 
718     56
719     61
720     88
740     66
741    123
Name: HP, Length: 70, dtype: int64

In [87]:
df.loc[(df['Type 1'] == 'Grass') & (df['HP'] > 100), ['Name', 'HP']]

,Name,HP
652,Amoonguss,114
741,Gogoat,123


In [88]:
df.loc[df['Type 1'] == 'Fire'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False


In [89]:
df.loc[df['HP'] > 100, ['Name', 'HP'] ].head(10)

,Name,HP
44,Jigglypuff,115
45,Wigglytuff,140
96,Muk,105
120,Rhydon,105
121,Chansey,250
123,Kangaskhan,105
124,KangaskhanMega Kangaskhan,105
142,Lapras,130
145,Vaporeon,130
155,Snorlax,160


Here are some pandas filtering challenges using a Pokémon dataset (assuming columns like `Name`, `Type 1`, `Type 2`, `HP`, `Attack`, `Defense`, `Speed`, `Legendary`, `Generation`).

Try writing the `.loc[]` filters yourself before checking the answers.

### Easy

1. Get all Grass Pokémon.
2. Get all Pokémon with HP greater than 100.
3. Get all Legendary Pokémon.
4. Get all Pokémon from Generation 1.
5. Get only the `Name` and `Attack` columns for Fire Pokémon.

```python
# Example format
df.loc[df['Type 1'] == 'Grass']
```



In [90]:
df.loc[(df['Type 1'] == 'Grass') | (df['Type 2'] == 'Grass')].sample(10).sort_values('Name' , ascending=True)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
110,102,Exeggcute,Grass,Psychic,60,40,80,60,45,40,1,False
659,598,Ferrothorn,Grass,Steel,74,94,131,54,116,20,5,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
617,556,Maractus,Grass,NaN,75,86,67,106,67,60,5,False
297,274,Nuzleaf,Grass,Dark,70,70,40,60,40,60,3,False
782,710,PumpkabooLarge Size,Ghost,Grass,54,66,70,44,55,46,6,False
536,479,RotomMow Rotom,Electric,Grass,50,65,107,105,107,86,4,False
550,492,ShayminLand Forme,Grass,NaN,100,100,100,100,100,100,4,True
203,188,Skiploom,Grass,Flying,55,45,50,45,65,80,2,False
432,387,Turtwig,Grass,NaN,55,68,64,45,55,31,4,False


In [91]:
df.loc[df['HP'] > 100].sample(4)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
594,534,Conkeldurr,Fighting,NaN,105,140,95,55,65,45,5,False
400,365,Walrein,Ice,Water,110,80,90,95,90,65,3,False
598,538,Throh,Fighting,NaN,120,100,85,30,85,45,5,False
559,500,Emboar,Fire,Fighting,110,123,65,100,65,65,5,False


In [92]:
df.loc[~df['Legendary']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
334,307,Meditite,Fighting,Psychic,30,40,55,40,55,60,3,False
723,655,Delphox,Fire,Psychic,75,69,72,114,100,104,6,False
282,260,Swampert,Water,Ground,100,110,90,85,90,60,3,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
560,501,Oshawott,Water,NaN,55,55,45,63,45,45,5,False


In [93]:
df.loc[df['Legendary']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
541,484,Palkia,Water,Dragon,90,120,100,150,120,100,4,True
711,646,KyuremBlack Kyurem,Dragon,Ice,125,170,100,120,90,95,5,True
792,716,Xerneas,Fairy,NaN,126,131,95,131,98,99,6,True
543,486,Regigigas,Normal,NaN,110,160,110,80,110,100,4,True
700,639,Terrakion,Rock,Fighting,91,129,90,72,90,108,5,True


In [94]:
df.loc[df['Generation'] == 1].sample(4)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
46,41,Zubat,Poison,Flying,40,45,35,30,40,55,1,False
130,121,Starmie,Water,Psychic,60,75,85,100,85,115,1,False
81,75,Graveler,Rock,Ground,55,95,115,45,45,35,1,False
30,25,Pikachu,Electric,NaN,35,55,40,50,50,90,1,False


In [95]:
df.loc[(df['Type 1'] == 'Fire') | (df['Type 2'] == 'Fire'),['Name', 'Attack']]

,Name,Attack
4,Charmander,52
5,Charmeleon,64
6,Charizard,84
7,CharizardMega Charizard X,130
8,CharizardMega Charizard Y,104
...,...,...
730,Fletchinder,73
731,Talonflame,81
735,Litleo,50
736,Pyroar,68


# Medium
### Get all Water Pokémon with HP above 80.
### Get all Pokémon whose Attack is greater than Defense.
### Get all Pokémon that are either Fire or Electric.
### Get all non-Legendary Pokémon.
### Get Pokémon with Speed between 80 and 120.

In [96]:
df.loc[(df['HP'] > 80) & ((df['Type 1'] == 'Water') | (df['Type 2'] == 'Water'))].sample(10).sort_values('Name', ascending=True)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
174,160,Feraligatr,Water,NaN,85,105,100,79,83,78,2,False
466,419,Floatzel,Water,NaN,85,105,55,85,50,115,4,False
470,423,Gastrodon,Water,Ground,111,83,68,92,82,39,4,False
140,130,Gyarados,Water,Flying,95,125,79,60,100,81,1,False
548,490,Manaphy,Water,NaN,100,100,100,100,100,100,4,False
67,62,Poliwrath,Water,Fighting,90,95,95,70,90,70,1,False
210,195,Quagsire,Water,Ground,95,85,85,65,65,35,2,False
85,79,Slowpoke,Water,Psychic,90,65,65,40,40,15,1,False
350,320,Wailmer,Water,NaN,130,70,35,70,35,60,3,False
351,321,Wailord,Water,NaN,170,90,45,90,45,60,3,False


In [97]:
df.loc[df['Attack'] > df['Defense']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
443,398,Staraptor,Normal,Flying,85,120,70,50,60,100,4,False
267,248,Tyranitar,Rock,Dark,100,134,110,95,100,61,2,False
665,604,Eelektross,Electric,NaN,85,115,80,105,80,50,5,False
769,699,Aurorus,Rock,Ice,123,77,72,99,92,58,6,False
767,697,Tyrantrum,Rock,Dragon,82,121,119,69,59,71,6,False


In [98]:
df.loc[
    (df['Type 1'].isin(['Fire', 'Electric'])) |
    (df['Type 2'].isin(['Fire', 'Electric']))
].sample(10).sort_values('Name')

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
435,390,Chimchar,Fire,NaN,44,58,44,58,44,61,4,False
723,655,Delphox,Fire,Psychic,75,69,72,114,100,104,6,False
337,309,Electrike,Electric,NaN,40,45,40,65,40,65,3,False
258,239,Elekid,Electric,NaN,45,63,37,65,55,95,2,False
764,694,Helioptile,Electric,Normal,44,38,33,61,43,70,6,False
135,126,Magmar,Fire,NaN,65,95,57,100,85,93,1,False
31,26,Raichu,Electric,NaN,60,90,55,90,80,110,1,False
276,255,Torchic,Fire,NaN,45,60,40,70,50,45,3,False
553,494,Victini,Psychic,Fire,100,100,100,100,100,100,5,True


In [99]:
df.loc[(df['Speed'] > 80) & (df['Speed'] < 120)]
# ==================OR========================
df.loc[df['Speed'].between(80, 120)].sample(10).reset_index()

,index,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,222,207,Gligar,Ground,Flying,65,75,105,35,65,85,2,False
1,43,38,Ninetales,Fire,NaN,73,76,75,81,100,100,1,False
2,152,141,Kabutops,Rock,Water,60,115,105,65,70,80,1,False
3,301,278,Wingull,Water,Flying,40,30,30,55,30,85,3,False
4,615,555,DarmanitanStandard Mode,Fire,NaN,105,140,55,30,55,95,5,False
5,544,487,GiratinaAltered Forme,Ghost,Dragon,150,100,120,100,120,90,4,True
6,131,122,Mr. Mime,Psychic,Fairy,40,45,65,100,120,90,1,False
7,698,637,Volcarona,Bug,Fire,85,60,65,135,105,100,5,False
8,211,196,Espeon,Psychic,NaN,65,65,60,130,95,110,2,False
9,381,349,Feebas,Water,NaN,20,15,20,10,55,80,3,False


# Hard
### Get Pokémon that are Grass AND Poison.
### Get Pokémon with HP > 100 and Attack > 100.
### Get Pokémon that are Legendary OR have Attack > 150.
### Get Pokémon that are not Water type.
### Get Pokémon whose name starts with "Char".

In [100]:
df.loc[
    df['Type 1'].isin(['Grass', 'Poison']) &
    df['Type 2'].isin(['Grass', 'Poison'])
]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False
49,44,Gloom,Grass,Poison,60,65,70,85,75,40,1,False
50,45,Vileplume,Grass,Poison,75,80,85,110,90,50,1,False
75,69,Bellsprout,Grass,Poison,50,75,35,70,30,40,1,False
76,70,Weepinbell,Grass,Poison,65,90,50,85,45,55,1,False
77,71,Victreebel,Grass,Poison,80,105,65,100,70,70,1,False


In [101]:
df.loc[
    (df['HP'] > 100) &
    (df['Attack'] > 100)
].sample(8)


,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
493,445,Garchomp,Dragon,Ground,108,130,95,80,85,102,4,False
96,89,Muk,Poison,NaN,105,105,75,65,100,50,1,False
524,473,Mamoswine,Ice,Ground,110,130,80,70,60,80,4,False
124,115,KangaskhanMega Kangaskhan,Normal,NaN,105,125,100,60,100,100,1,False
321,297,Hariyama,Fighting,NaN,144,120,60,40,60,50,3,False
120,112,Rhydon,Ground,Rock,105,130,120,45,45,40,1,False
792,716,Xerneas,Fairy,NaN,126,131,95,131,98,99,6,True
543,486,Regigigas,Normal,NaN,110,160,110,80,110,100,4,True


# filtering from the sub element of a  value

In [102]:
df.loc[df['Name'].str.startswith('Char')]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False


In [103]:
df.loc[df['Name'].str.contains('Mega')].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
87,80,SlowbroMega Slowbro,Water,Psychic,95,75,180,130,80,30,1,False
283,260,SwampertMega Swampert,Water,Ground,100,150,110,95,110,70,3,False
426,384,RayquazaMega Rayquaza,Dragon,Flying,105,180,100,180,100,115,3,True
19,15,BeedrillMega Beedrill,Bug,Poison,65,150,40,15,80,145,1,False
248,229,HoundoomMega Houndoom,Dark,Fire,75,90,90,140,90,115,2,False


# SAFER filtering with .loc and additional parameters


| Goal                    | Fix          |
| ----------------------- | ------------ |
| Case-insensitive search | `case=False` |
| Avoid NaN errors        | `na=False`   |

    case=False → ignores uppercase/lowercase differences || matches: Mega, mega, MEGA, mEGa

    na=False → treats NaN as “not matching”

If you do:
```python
df['Name'].str.contains('Mega')
```
and there are NaN values in Name, pandas returns:

    NaN
    True
    False
    NaN

It does not automatically drop them.

In [104]:
df.loc[df['Name'].str.contains('Mega',case=False, na=False)].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
196,181,AmpharosMega Ampharos,Electric,Dragon,90,95,105,165,110,45,2,False
476,428,LopunnyMega Lopunny,Normal,Fighting,65,136,94,54,96,135,4,False
591,531,AudinoMega Audino,Normal,Fairy,103,60,126,80,126,50,5,False
248,229,HoundoomMega Houndoom,Dark,Fire,75,90,90,140,90,115,2,False
796,719,DiancieMega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True


In [105]:
import re

df.loc[df['Type 1'].str.contains('fire|grass', flags=re.I, regex=True)].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
111,103,Exeggutor,Grass,Psychic,95,95,85,125,65,55,1,False
570,511,Pansage,Grass,NaN,50,53,48,53,48,64,5,False
452,407,Roserade,Grass,Poison,60,70,65,125,105,90,4,False
701,640,Virizion,Grass,Fighting,91,90,72,90,129,108,5,True
720,652,Chesnaught,Grass,Fighting,88,107,122,74,75,64,6,False


In [110]:
df.loc[df['Name'].str.endswith('saur', na=False)]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False


In [111]:
df.loc[df['Name'].str.contains(r'saur$', case=False, na=False)]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
